lib/class-wp-rest-url-details-controller.php

```diff
@@ -0,0 +1,201 @@
+<?php
+/**
+ * REST API: WP_REST_URL_Details_Controller class
+ *
+ * @package Gutenberg
+ */
+
+/**
+ * Controller which provides REST endpoint for retrieving information
+ * from a remote site's HTML response.
+ *
+ * @since 5.?.0
+ *
+ * @see WP_REST_Controller
+ */
+class WP_REST_URL_Details_Controller extends WP_REST_Controller {
+
+	/**
+	 * Constructs the controller.
+	 *
+	 * @access public
+	 */
+	public function __construct() {
+		$this->namespace = '__experimental';
+		$this->rest_base = 'url-details';
+	}
+
+	/**
+	 * Registers the necessary REST API routes.
+	 *
+	 * @access public
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'parse_url_details' ),
+					'args'                => array(
+						'url' => array(
+							'required'          => true,
+							'description'       => __( 'The URL to process.', 'gutenberg' ),
+							'validate_callback' => 'wp_http_validate_url',
+							'sanitize_callback' => 'esc_url_raw',
+							'type'              => 'string',
+							'format'            => 'uri',
+						),
+					),
+					'permission_callback' => array( $this, 'permissions_check' ),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Retrieves the contents of the <title> tag from the HTML
+	 * response.
+	 *
+	 * @access public
+	 * @param WP_REST_REQUEST $request Full details about the request.
+	 * @return String|WP_Error The title text or an error.
+	 */
+	public function parse_url_details( $request ) {
+
+		$url = untrailingslashit( $request->get_param( 'url' ) );
+
+		if ( empty( $url ) ) {
+			return new WP_Error( 'rest_invalid_url', __( 'Invalid URL', 'gutenberg' ), array( 'status' => 404 ) );
+		}
+
+		// Transient per URL.
+		$cache_key = $this->build_cache_key_for_url( $url );
+
+		// Attempt to retrieve cached response.
+		$data = $this->get_cache( $cache_key );
+
+		// Return cache if valid data.
+		if ( false && ! is_wp_error( $data ) && ! empty( $data ) ) {
```

Comment:
```suggestion
		if ( ! empty( $data ) ) {
```

* stray `false` (this condition is never met)
* errors are never cached (and shouldn't be!)

Reply:
> stray false (this condition is never met)

About to replace this with an `apply_filters()` call.

Reply:
> errors are never cached (and shouldn't be!)

Guess I'm being overly cautious.